presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +702,115 @@ public void prepareFinal()
             });
         }
     }
+
+    private static class SpillableOrderingGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final OrderingGroupedAccumulator orderingGroupedAccumulator;
+        private final ObjectBigArray<RowBlockBuilder> intermediateState;
+        private final List<Type> spillingTypes;
+
+        public SpillableOrderingGroupedAccumulator(
+                GroupedAccumulator accumulator,
+                List<Type> aggregationSourceTypes,
+                List<Integer> orderByChannels,
+                List<SortOrder> orderings,
+                PagesIndex.Factory pagesIndexFactory)
+        {
+            this.orderingGroupedAccumulator = new OrderingGroupedAccumulator(accumulator, aggregationSourceTypes, orderByChannels, orderings, pagesIndexFactory);
+            this.spillingTypes = aggregationSourceTypes;
+            this.intermediateState = new ObjectBigArray<>();
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return orderingGroupedAccumulator.getEstimatedSize() + intermediateState.sizeOf();
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return orderingGroupedAccumulator.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            return new ArrayType(RowType.anonymous(spillingTypes));
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        {
+            for (int pos = 0; pos < groupIdsBlock.getPositionCount(); pos++) {
+                long groupId = groupIdsBlock.getGroupId(pos);
+                intermediateState.ensureCapacity(groupId);
+                RowBlockBuilder rowBlockBuilder = intermediateState.get(groupId);
+                if (rowBlockBuilder == null) {
+                    rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIdsBlock.getGroupCount());
+                }
+
+                BlockBuilder output = rowBlockBuilder.beginBlockEntry();
+                for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                    spillingTypes.get(channel).appendTo(page.getBlock(channel), pos, output);
+                }
+                rowBlockBuilder.closeEntry();
+
+                intermediateState.set(groupId, rowBlockBuilder);
+            }
+
+            orderingGroupedAccumulator.addInput(groupIdsBlock, page);
```

Comment:
This should probably happen in prepareFinal() but unsure how to do this without using twice as much memory.